File: Recursion/sum2/Solution.py

# Problem code
def combinationSum2(candidates, target):
    result = []
    backtrack(0, sorted(candidates), target, result, [])
    return result

def backtrack(start_index, candidates, target, result, temp):
    if start_index > len(candidates) or target < 0:
        return

    if target == 0:
        result.append(temp.copy())
        return

    for i in range(start_index, len(candidates)):
        candidate = candidates[i]
        if candidate > target:
            return

        if i > start_index and candidates[i] == candidates[i - 1]:
            continue

        # use it
        target -= candidate
        temp.append(candidate)

        # backtrack
        backtrack(i + 1, candidates, target, result, temp)

        # unuse it
        target += candidate
        temp.pop()

File: Recursion/sum2/Solution.swift

// Problem code
func combinationSum2(_ candidates: [Int], _ target: Int) -> [[Int]] {
	var result = [[Int]]()
	var temp = [Int]()
	backtrack(0, candidates.sorted(), target, &result, &temp)
	return result
}

func backtrack(_ startIndex: Int, _ candidates: [Int], _ target: Int, _ result: inout [[Int]], _ temp: inout [Int]) {
	if startIndex > candidates.count || target < 0 {
		return
	}
	
	if target == 0 {
		result.append(temp)
		return
	}
	
	var sum = target
	for i in startIndex..<candidates.count {
		if i > startIndex && candidates[i] == candidates[i - 1] {
			continue
		}
		let candidate = candidates[i]
		if candidate > sum {
			return
		}
		
		// use it
		sum -= candidate
		temp.append(candidate)
		
		// backtrack
		backtrack(i + 1, candidates, sum, &result, &temp)
		
		// unuse it
		sum += candidate
		_ = temp.popLast()
	}
}
